src/tensors/gpu/integer_tools.cu

```diff
@@ -580,6 +720,60 @@ namespace integer {
         findMaxMinAndQuantMult<<<1,1>>>(input_gpu, resMaxIdx - 1, resMinIdx - 1, output_gpu); //FUCK YOU FORTRAN INDEXING
     }
 
+    // @TODO some bug in this method
+    __global__ void maxAbsQuantMultKernel(const half *input_gpu, float *output_gpu, size_t size) {
+      int idx = blockDim.x * blockIdx.x + threadIdx.x;
+      __shared__ float sdata[256];
+
+      if(idx < size) {
+        sdata[threadIdx.x] = abs(__half2float(input_gpu[idx]));
+        __syncthreads();
+
+        for(int stride = blockDim.x >> 1; stride > 0; stride >>= 1) {
```

Comment:
What's happening here? Why is blockDim.x being shifted right?

Reply:
Did you do the nice rewrite for me?

Reply:
That's the reduction. The idea is to use the multiple thread blocks in a GPU to reduce a small portion of the array. Because cublasIsmax and cublasismin don't support half. I thought it might be better write with a half kernel. (But it's quite hard to get it right)

Reply:
Oh, right, you're getting half of the problem at each step. I see